src/node/interfaces.cpp

```diff
@@ -870,6 +870,30 @@ class MinerImpl : public Mining
         return BlockRef{tip->GetBlockHash(), tip->nHeight};
     }
 
+    BlockRef waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
+    {
+        // Interrupt check interval
+        const MillisecondsDouble tick{1000};
+        auto now{std::chrono::steady_clock::now()};
+        auto deadline = now + timeout;
+        // std::chrono does not check against overflow
+        if (deadline < now) deadline = std::chrono::steady_clock::time_point::max();
+        {
+            WAIT_LOCK(g_best_block_mutex, lock);
+            while ((g_best_block == uint256() || g_best_block == current_tip) && !chainman().m_interrupt) {
+                now = std::chrono::steady_clock::now();
+                if (now >= deadline) break;
+                const MillisecondsDouble remaining{deadline - now};
+                const auto interval{std::min(remaining, tick)};
+                g_best_block_cv.wait_until(lock, now + interval);
```

Comment:
In commit "Add waitTipChanged to Mining interface" (d1c197fbbad86e3f966d9bf9abe4dd0aa76bc83d)

Not important but I think it would be clearer to condense these 3 lines to just:

```c++
g_best_block_cv.wait_until(lock, std::min(deadline, now + tick));
```

I don't think it helps to introduce 2 extra variables and subtract `now` then add `now` back.

Reply:
Indeed that's nicer.